File: algorithms/0002.add-two-numbers/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int val = l1.val + l2.val;
        ListNode res = new ListNode(val % 10);
        ListNode current_res = res;

        int over = (val >= 10) ? 1 : 0;
        l1 = l1.next;
        l2 = l2.next;

        while (over > 0 || l1 != null || l2 != null) {
            val = over;
            val += (null == l1) ? 0 : l1.val;
            val += (null == l2) ? 0 : l2.val;

            current_res.next = new ListNode(val % 10);

            over = (val >= 10) ? 1 : 0;
            l1 = (null == l1) ? l1 : l1.next;
            l2 = (null == l2) ? l2 : l2.next;
            current_res = current_res.next;
        }

        return res;
    }
}

File: algorithms/0002.add-two-numbers/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	val := l1.Val + l2.Val
	res := &ListNode{Val: val % 10, Next: nil}
	current := res

	over := 0
	if val >= 10 {
		over = 1
	}

	l1 = l1.Next
	l2 = l2.Next

	for over > 0 || l1 != nil || l2 != nil {
		val = over
		if l1 != nil {
			val += l1.Val
		}
		if l2 != nil {
			val += l2.Val
		}

		current.Next = &ListNode{Val: val % 10, Next: nil}

		over = 0
		if val >= 10 {
			over = 1
		}

		if l1 != nil {
			l1 = l1.Next
		}
		if l2 != nil {
			l2 = l2.Next
		}

		current = current.Next
	}

	return res
}
